### DISCUSSION TITLE
Any suggestions for aligning a set of linear equations?

### DISCUSSION THREAD

**Discussion by jason-s** [KaTeX/KaTeX#2552] [General]
What's the best way to align a set of linear equations in KaTeX? It doesn't support `align*` or `eqnarray`.

I can't seem to use `\begin{aligned}` to get the terms lined up nicely. The best I can do is this kludge:

```
\begin{matrix}X_\alpha \\ \\ X_\beta\end{matrix}
\begin{matrix}{}={} \\ \\{}={}\end{matrix}
\begin{matrix}\frac{2}{3}X_A \\ \\{}\end{matrix}
\begin{matrix}{}-{} \\ \\ {}\end{matrix}
\begin{matrix}\frac{1}{3}X_B \\ \\ \frac{1}{\sqrt{3}}X_B\end{matrix}
\begin{matrix}{}-{} \\ \\ {}-{}\end{matrix}
\begin{matrix}\frac{1}{3}X_C \\ \\{}\frac{1}{\sqrt{3}}X_C\end{matrix}
```

![image](https://user-images.githubusercontent.com/738893/94494179-645c4e00-01a3-11eb-8ad1-403344d35fa0.png)

[This SO question](https://stackoverflow.com/questions/48240982/how-do-you-align-a-system-of-equations/48266309) suggests using `systeme` but I'm guessing that's not supported in KaTeX.

**Comment by jason-s** [comment]
oh, looks like I can use `array` as in https://tex.stackexchange.com/a/328771/46076

```
\begin{array}{rcrcrc}
X_\alpha&=&\frac{2}{3}X_A&-&\frac{1}{3}X_B &-&\frac{1}{3}X_C \\
X_\beta&=& & &\frac{1}{\sqrt{3}}X_B &-&\frac{1}{\sqrt{3}}X_C \\
\end{array}
```

![image](https://user-images.githubusercontent.com/738893/94494469-1e53ba00-01a4-11eb-9e90-f0335bd2cdd2.png)

or 

```
\begin{array}{rcrrr}
X_\alpha&=&\frac{2}{3}X_A&-\frac{1}{3}X_B &-\frac{1}{3}X_C \\
X_\beta&=&  &\frac{1}{\sqrt{3}}X_B &-\frac{1}{\sqrt{3}}X_C \\
\end{array}
```

![image](https://user-images.githubusercontent.com/738893/94494549-4d6a2b80-01a4-11eb-961d-da05961bb997.png)

but both have yucky-looking horizontal and vertical spacing.

**Comment by edemaine** [comment]
@jason-s If `align*` is what you want, KaTeX supports the nearly identical `aligned` environment. I don't know how to use that to align linear equations well though. What you're essentially asking for is `\arraycolsep` support, which is on the to-do list but not currently supported.

**Reply by jason-s** [reply]
I'm familiar with the use of `aligned` and it works very well for equations needing only one alignment point, as in

```
\begin{aligned}
X_\alpha&=\frac{2}{3}X_A-\frac{1}{3}X_B -\frac{1}{3}X_C \\
X_\beta&=\frac{1}{\sqrt{3}}X_B -\frac{1}{\sqrt{3}}X_C
\end{aligned}
```

![image](https://user-images.githubusercontent.com/738893/94496612-b6a06d80-01a9-11eb-9538-8a64ad6b305b.png)

but I would really like to align the terms so that the X_A, X_B, and X_C line up.

This seems like a common enough use case for systems of linear equations that KaTeX should support it somehow... just not sure how to get it the way you can in LaTeX. (not sure MathJax does either, however.)

**Reply by edemaine** [reply]
Like I said, what's missing is `\arraycolsep` support. Then your examples above will look great. And this is how I'd do it in LaTeX.

**Reply by jason-s** [reply]
Ah, ok, I'll wait patiently :-)

**Comment by jason-s** [comment]
Ooh! I just noticed `alignedat` in https://katex.org/docs/supported.html -- now to try it....

hmmm, this is a somewhat confusing feature:

```
\begin{alignedat}{4}
   X_\alpha &=\frac{2}{3}X_A &-& \frac{1}{3}X_B &- \frac{1}{3}&X_C \\
   X_\beta &= && \frac{1}{\sqrt{3}}X_B &- \frac{1}{\sqrt{3}}&X_C
\end{alignedat}
```

![image](https://user-images.githubusercontent.com/738893/94598601-0a5f9500-0244-11eb-8d1b-cf1e59aaf255.png)

OK one more attempt with a fudge factor using manual spacing to help center the coefficients:

```
\begin{alignedat}{4}
   X_\alpha &=\tfrac{2}{3}X_A &-& \;\tfrac{1}{3}&X_B &- \enspace\tfrac{1}{3}&X_C \\
   X_\beta &= && \tfrac{1}{\sqrt{3}}&X_B &- \;\tfrac{1}{\sqrt{3}}&X_C
\end{alignedat}
```

![image](https://user-images.githubusercontent.com/738893/94601379-4268d700-0248-11eb-9e27-6fc3e63d115c.png)

ok minor tweaks:

```
\begin{alignedat}{4}
   X_\alpha &=\tfrac{2}{3}X_A &{}-{}& \;\tfrac{1}{3}&X_B &{}-{} \;\tfrac{1}{3}&X_C \\
   X_\beta &= && \tfrac{1}{\sqrt{3}}&X_B &- \tfrac{1}{\sqrt{3}}&X_C
\end{alignedat}
```

![image](https://user-images.githubusercontent.com/738893/94602290-83adb680-0249-11eb-83e5-a8a892d1fe1e.png)

**Reply by edemaine** [reply]
Maybe see https://mirrors.rit.edu/CTAN/macros/latex/required/amsmath/amsldoc.pdf#page=8 for how to use align(ed)at.

**Reply by jason-s** [reply]
yeah, I looked there and I sort of get it... what's the difference between `alignat` and `alignedat`?

**Reply by edemaine** [reply]
`alignat` is how you'd enter math mode from vertical text mode, while `alignedat` is what you use when you're already in math mode.  KaTeX only supports math mode, so it only supports the latter.

**Reply by jason-s** [reply]
ah -- ok that makes sense, thanks!

**Reply by ronkok** [reply]
Another explanation for `alignat` can be found at: https://www.onemathematicalcat.org/MathJaxDocumentation/TeXSyntax.htm#alignat

**Reply by snorky22** [reply]
It is all too bad that eqnarray is not supported because it is featured in the original LaTeX book by  Leslie Lamport and used by quite a few people that I know. It is a classic.

**Reply by snorky22** [reply]
I have tried it and the  alignedat environment is definitively not a substitute for the  eqnarray environment
In eqnarray, the first part is flushed right, the second part is centered and the third part is flushed left.
You do not have that at all in alignedat.
Too bad, I will stick to Mathjax.
